wallet: re-activate  "AmountWithFeeExceedsBalance" error

furszy opened the pull request:
During a talk with theStack, it was noted that the `AmountWithFeeExceedsBalance` error inside `WalletModel::prepareTransaction` is never thrown.

This is because `createTransaction` does not retrieve the fee when the process fails due to insufficient funds since #20640. The fee return arg is set only at the end of the process, when the transaction is successfully created. Therefore, if the transaction creation fails, the fee is not available inside `WalletModel::prepareTransaction` to trigger the `AmountWithFeeExceedsBalance` error.

This PR re-implements the feature inside `createTransaction` and adds test coverage for it.

luke-jr commented:
Seems like an awful lot of refactoring and optimisations tied into what is in essence a bugfix.

Maybe split them up?

NawaraTafa commented:
مساء الخير

furszy commented:
> Seems like an awful lot of refactoring and optimisations tied into what is in essence a bugfix.
> Maybe split them up?

Hey Luke, only the last three commits count for this work. It's mentioned in the PR description to not generate confusion: This was built on top of #25005 because needed the structure introduced in 5b6124da to implement this flow properly. (where "properly" means not doing an ugly workaround and place the new code on top of a cleaner structure)

Now that #25005 got merged, will rebase it and get rid of most of the extra commits.

luke-jr commented:
Looks like this works around a regression introduced by #20640

Might be better to make `CreateTransaction` return a (non-optional) `CreatedTransactionResult` instead?

furszy commented:
Hey @luke-jr thanks for the review, was tackling other PRs before moving back to this one.

> Looks like this works around a regression introduced by #20640
> 
> Might be better to make `CreateTransaction` return a (non-optional) `CreatedTransactionResult` instead?

As we are now returning an `std::variant` wrapper, that change might not worth it (it would mean re-introduce the string error ref argument and change all the function callers, and the many return statements of `CreateTransaction` and `CreateTransactionInternal`).

A possible elegant solution going into the "always return information" path would be to return an object that wraps the transaction creation process information (with the fee, the used coin selection algorithm, change output, etc..) in the error field of the returned `std::variant` as well. But.. for that, we need #25601 which introduces the generic error functionality.

But aside from that (which I think that would be a good long term goal), maybe might worth to continue with the PR as is now, primarily because it's unifying the errors that we throw on the GUI and on the RPC server.

Still, have to say that I would like to be able to move this kind of errors from the wallet sources to an upper class like the wallet interface. So GUI and RPC receive the same errors and the wallet isn't in charge of describe this type of errors for the user.

furszy reviewed (commented):
Based on yesterday's wallet meeting, have been thinking about 032e5449 further and the possibility to move the error to `SelectCoins` instead of placing it inside `CreateTransaction`.

The commit, as is now, it does not cover the "fee exceeds balance" error in a 100%. It's only covering the "fee exceeds balance" if the selection fails for the not-inputs fees, it is not contemplating the inputs fees (this is not different to what we were doing before #20640).

So, need to move the error to `SelectCoins`, and add a mechanism to keep track of the inputs fee total amount (data that we don't have in `CreateTransaction` if `SelectCoins` fail).

But.. ideally, before implementing this changes, would love to get #25806 merged as it simplifies the whole coin selection process greatly (which depends on #25685, which is ready to go).

NawaraTafa commented:
لم افهم مقصود او معني الرساله!!

في سبت، 8 أكتوبر، 2022 في 6:28 م، كتب Matias Furszyfer <
***@***.***>:

> ***@***.**** commented on this pull request.
>
> Based on yesterday's wallet meeting, have been thinking about 032e544
> <https://github.com/bitcoin/bitcoin/commit/032e5449e53f40b6de1e5858abeab544ea848614>
> further and the possibility to move the error to SelectCoins instead of
> placing it inside CreateTransaction.
>
> The commit, as is now, it does not cover the "fee exceeds balance" error
> in a 100%. It's only covering the "fee exceeds balance" if the selection
> fails for the not-inputs fees, it is not contemplating the inputs fees
> (this is not different to what we were doing before #20640
> <https://github.com/bitcoin/bitcoin/pull/20640>).
>
> So, need to move the error to SelectCoins, and add a mechanism to keep
> track of the inputs fee total amount (data that we don't have in
> CreateTransaction if SelectCoins fail).
>
> But.. ideally, before implementing this changes, would love to get #25806
> <https://github.com/bitcoin/bitcoin/pull/25806> merged as it simplifies
> the whole coin selection process greatly (which depends on #25685
> <https://github.com/bitcoin/bitcoin/pull/25685>, which is ready to go).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/25269#pullrequestreview-1135240825>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZMEHAUQZ7SCS3NKM3GNSUTWCGOJXANCNFSM5XVEIO3A>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
Noura khaled

fanquake commented:
> But.. ideally, before implementing this changes, would love to get https://github.com/bitcoin/bitcoin/pull/25806 merged
> (which depends on https://github.com/bitcoin/bitcoin/pull/25685, which is ready to go).

25685 has been merged, and it looks like 25806 has just been rebased. How about marking this PR draft for now, given the dependency on 25806, as well as adding a note at the top of the PR description, indicating to reviewers that they should review 25806 first.

furszy committed:
wallet: introduce "tx amount exceeds balance when fees are included" error

This was previously implemented at the GUI level but has been broken since #20640

murchandamus reviewed (commented):
Concept ACK

furszy commented on src/wallet/spend.cpp:
```diff
@@ -1131,7 +1131,19 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
+        // Check if we have enough balance but cannot cover the fees
+        CAmount available_balance = preset_inputs.total_amount + available_coins.GetTotalAmount();
+        if (available_balance >= recipients_sum) {
+            CAmount available_effective_balance = preset_inputs.total_amount + available_coins.GetEffectiveTotalAmount().value_or(available_coins.GetTotalAmount());
```
> It seems that both in lines 1137 and 1139 the preset_inputs are counted with their full amount rather than their effective amount. Should the `preset_inputs` not also be reduced to their effective amount in line 1139?

Thats because `preset_inputs.total_amount` could either be the full amount or the effective one (see `PreSelectedInputs`). I didn't want to introduce more changes within this PR but we should split the amounts (probably replacing `PreSelectedinputs` usage for `CoinsResult`).

murchandamus commented on src/wallet/spend.cpp:
```diff
@@ -1161,7 +1161,19 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
+        if (!err.empty()) return util::Error{err};
+
+        // Check if we have enough balance but cannot cover the fees
+        CAmount available_balance = preset_inputs.total_amount + available_coins.GetTotalAmount();
```
Note to reviewers: `preset_inputs.total_amount` will be the sum of the outputs’ effective values if "subtract fee from outputs" is disabled, but will be the nominal amount if SFFO is active. See [`src/wallet/spend.h:164–172`](https://github.com/bitcoin/bitcoin/blob/8cefff322cc6689e50074c93cb9d52412d614d13/src/wallet/spend.h#L164)

murchandamus commented on src/wallet/spend.cpp:
```diff
@@ -1161,7 +1161,19 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
+        // Check if we have enough balance but cannot cover the fees
+        CAmount available_balance = preset_inputs.total_amount + available_coins.GetTotalAmount();
+        if (available_balance >= recipients_sum) {
+            CAmount available_effective_balance = preset_inputs.total_amount + available_coins.GetEffectiveTotalAmount().value_or(available_coins.GetTotalAmount());
```
I’m wondering whether there is a bug here. It seems to me that this is assuming that the effective value is only available when SFFO is not being used, but `COutput.effective_value` is set and aggregated whenever the COutputs have a feerate. Does this not explicitly need to handle SFFO? I.e., maybe this needs to be something along the lines of:

```suggestion
            CAmount available_coins_total_amount = coin_selection_params.m_subtract_fee_outputs ? available_coins.GetTotalAmount() : available_coins.GetEffectiveTotalAmount();
            CAmount available_effective_balance = preset_inputs.total_amount +available_coins_total_amount;
```

achow101 replied:
Since fees are not included when SFFO, I don't think this is reachable as should `available_balance` is always greater than `recipients_sum` in that case, so I don't think this needs any special SFFO treatment. Furthermore, if it were possible to reach this with SFFO, I think the error message would be confusing and it would actually be better to return "Insufficent Funds", so this whole check could probably be guarded by an `if (m_subtract_fee_outputs)` to skip it when SFFO is on.

furszy replied:
ha.. true. It seems I'm not putting much brain power here.. updated per feedback. Thanks.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25269#discussion_r2115800834

Thinking about this more I can see why SFFO should make the "amount exceeds your balance when fees are included" condition harder to trigger than when SFFO is not used, because when SFFO is enabled the wallet is able to reduce the amounts recipients receive to cover any fees, whereas without SFFO the amount received needs to be fixed. But I'd still expect the `(available_balance >= recipients_sum && available_effective_balance < selection_target)` condition to be able to trigger when SFFO is used if there are coins are which are not economical to send. Also even if there is other code making this condition difficult or impossible to trigger when SFFO is used, I still don't see a reason this code should be checking for SFFO specially. Even with SFFO, an error message that mentions fees and includes the attempted fee would still be more descriptive and helpful than the generic "insufficient funds" message.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25269#discussion_r2115800834

Update: think I finally figured this out. In the SFFO case the fee exceeds balance error can still happen but the `available_balance >= recipients_sum` condition will not detect it. So I think the change I would suggest would just be dropping the `m_subtract_fee_outputs` code and writing a more descriptive comment:

```diff
--- a/src/wallet/spend.cpp
+++ b/src/wallet/spend.cpp
@@ -1164,12 +1164,19 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
 
         // Check if we have enough balance but cannot cover the fees
         CAmount available_balance = preset_inputs.total_amount + available_coins.GetTotalAmount();
-        // Note: Since SFFO reduces existing outputs to cover fees, recipients_sum should already be
-        // strictly less than available_balance when enabled. Still, as a sanity-check, skip when SFFO is enabled.
-        if (available_balance >= recipients_sum && !coin_selection_params.m_subtract_fee_outputs) {
+        if (available_balance >= recipients_sum) {
             CAmount available_effective_balance = preset_inputs.total_amount + *available_coins.GetEffectiveTotalAmount();
             if (available_effective_balance < selection_target) {
-                return util::Error{_("The total transaction amount exceeds your balance when fees are included")};
+                // Note: this fees exceeds balance error will not currently
+                // trigger when SFFO (m_subtract_fee_outputs) is enabled because
+                // the effective balance will equal the original balance in this
+                // case. This is not ideal because it fails to provide the most
+                // descriptive error message, but it's not worth making this
+                // code more complex by checking the sizes of outputs fees can
+                // be subtracted from. Additionally, fees exceeding the balance
+                // should happen more rarely in the SFFO case given the ability
+                // to take fees from outputs.
+                return util::Error{strprintf(_("The total exceeds your balance when the %s transaction fee is included."), FormatMoney(selection_target - recipients_sum))};
             }
         }
 
```

murchandamus commented on test/functional/wallet_fundrawtransaction.py:
```diff
@@ -1522,5 +1524,20 @@ def test_input_confs_control(self):
+
+        rawtx = wallet.createrawtransaction(inputs=[], outputs=[{self.nodes[0].getnewaddress(): 0.3}])
+        expected_err_msg = "The total transaction amount exceeds your balance when fees are included"
+        assert_raises_rpc_error(-4, expected_err_msg, wallet.fundrawtransaction, rawtx)
```
Could this test case be extended by attempting the same transaction with SFFO and showing that it succeeds?

murchandamus reviewed (commented):
I’m wondering whether this is correct when SFFO is active.

furszy committed:
wallet: fix, make 'total_effective_amount' optional actually optional

furszy commented:
Updated per feedback. Thanks murchandamus!

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -1160,7 +1160,21 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
+        CAmount available_balance = preset_inputs.total_amount + available_coins.GetTotalAmount();
+        // Note: Since SFFO reduces existing outputs to cover fees, recipients_sum should already be
+        // strictly less than available_balance when enabled. Still, as a sanity-check, skip when SFFO is enabled.
+        if (available_balance >= recipients_sum && !coin_selection_params.m_subtract_fee_outputs) {
```
In 98b7f1061ff104cdffebe1c2968c9f829d1a72f9 "wallet: introduce "tx amount exceeds balance when fees are included" error"

I'm pretty sure that `!coin_selection_params.m_subtract_fee_outputs` is not necessary as it should not be possible for `SelectCoins` to fail but `available_balance >= recipients_sum` with SFFO.

Perhaps this could be made into an `Assume` inside the `if`?

ryanofsky reviewed (commented):
Code review ACK 98b7f1061ff104cdffebe1c2968c9f829d1a72f9 for the second commit, but I am unsure about the first commit (see below). I started reviewing this because https://github.com/bitcoin-core/gui/pull/807 depends on it, but this seems like a nice change independently that provides a clearer error message and fixes a GUI feature that was accidentally broken in #20640.

stratospher commented on src/wallet/spend.h:
```diff
@@ -59,13 +59,13 @@ struct CoinsResult {
     CAmount total_amount{0};
     /** Sum of all available coins effective value (each output value minus fees required to spend it) */
-    std::optional<CAmount> total_effective_amount{0};
+    std::optional<CAmount> total_effective_amount;
```
> I wonder if this change requires updates to any code that might be assuming total_effective_amount is not null like

I don't think it's possible to reach here without `*total_effective_amount` already existing. but added a check for whether the optional value exists just in case.

looks like there's only 2 places where `CoinsResult::total_effective_amount` is used - in `CoinsResult::Erase` and `CoinsResult::Add`. it's accessed safely in `Add()` and now added an extra check in `Erase()`. only usage for `Erase()` in the codebase seems to be in the tests.

> Also it's not really clear what the connection is between this change in the rest of the PR. It would help to mention some motivation in the commit description.

it was just a clean-up along the way noticed during review - https://github.com/bitcoin/bitcoin/pull/25269#discussion_r2107038658. added that motivation to commit description in https://github.com/bitcoin/bitcoin/pull/34299.
